### PR TITLE
VRF subscription mock

### PR DESCRIPTION
**Pull request by thedriftofwords**
Copy of vrf-more branch from #1931 

- https://documentation-git-fork-thedriftofwords-vrf-09248c-chainlinklabs.vercel.app/vrf/v2-5/subscription/test-locally

### CODE REVIEW

File: public/samples/VRF/mock/VRFv2_5Consumer.sol [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    // The default is 3, but you can set this higher.
+    uint16 constant REQUEST_CONFIRMATIONS = 3;
+
+    // For this example, retrieve 2 random values in one request.
+    // Cannot exceed VRFCoordinatorV2_5.MAX_NUM_WORDS.
+    uint32 constant NUM_WORDS = 2;
+
+    uint256[] public s_randomWords;
+    uint256 public s_requestId;
+    address s_owner;
```
[the remark this answers is not in the document]
Comment (khadni) [reply public/samples/VRF/mock/VRFv2_5Consumer.sol:44]: The consumer contract already inherits the storage variable `owner` from `ConfirmedOwnerWithProposal`, it is not necessary to redefine it.

File: public/samples/VRF/mock/VRFv2_5Consumer.sol [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     * @param vrfCoordinator - coordinator, check https://docs.chain.link/vrf/v2-5/supported-networks
+     * @param keyHash - the gas lane to use, which specifies the maximum gas price to bump to
+     */
+    constructor(
+        uint256 subscriptionId,
+        address vrfCoordinator,
+        bytes32 keyHash
+    ) VRFConsumerBaseV2Plus(vrfCoordinator) {
+        s_keyHash = keyHash;
+        s_owner = msg.sender;
```
Comment (khadni) [inline public/samples/VRF/mock/VRFv2_5Consumer.sol:61]:
```suggestion

```

Comment (khadni) [reply public/samples/VRF/mock/VRFv2_5Consumer.sol:61]: same as above

**Review by khadni — CHANGES_REQUESTED** [review]

File: public/samples/VRF/mock/VRFv2_5Consumer.sol [hunk 3]
```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: MIT
+// An example of a consumer contract that relies on a subscription for funding.
+pragma solidity ^0.8.19;
```
[the remark this answers is not in the document]
Comment (thedriftofwords) [reply public/samples/VRF/mock/VRFv2_5Consumer.sol]: Yes, thanks! Updated in [b4c5cdd](https://github.com/smartcontractkit/documentation/pull/1946/commits/b4c5cdd98c1448d204f4521bb7b60610def9446d)

File: src/content/vrf/v2-5/subscription/test-locally.mdx [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+   <ClickToZoom src="/images/vrf/v2-5/mock/manual-fulfill-request.png" />
+
+1. Click on `transact`.
+
+### Check the results
+
+1. Under _Deployed Contracts_, open the functions list of your deployed _RandomNumberConsumerV2_5_ contract.
+
+1. Each time you make a VRF request, your consumer contract requests two random words. After the request is fulfilled, the two random words are stored in the `s_randomWords` array. You can check the stored random words by reading the two first indexes of the `s_randomWords` array. To do so, click on the _s_randomWords_ function and:
```
Comment (dwightjl) [inline src/content/vrf/v2-5/subscription/test-locally.mdx:156]:
```suggestion
1. For each VRF request, your consumer contract requests two random words. After a request is fulfilled, the two random words are stored in the `s_randomWords` array. You can check the stored random words by reading the two first indexes of the `s_randomWords` array. To do so, click the _s_randomWords_ function and:
```

File: src/content/vrf/v2-5/subscription/test-locally.mdx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+      {" "}
+
+      <ClickToZoom src="/images/vrf/v2-5/mock/show-random-word.png" />
+
+   1. You can read the second random word in a similar way: fill in the index with _1_ then click on _call_ to display the second random word.
+
+## Next steps
+
+This guide demonstrated how to test a VRF v2.5 consumer contract on your local blockchain. We made the guide on RemixIDE for learning purposes, but you can reuse the same [testing logic](#testing-logic) on another development environment, such as Hardhat. For example, read the Hardhat Starter Kit [RandomNumberConsumer unit tests](https://github.com/smartcontractkit/hardhat-starter-kit/blob/main/test/unit/RandomNumberConsumer.spec.js).
```
Comment (dwightjl) [inline src/content/vrf/v2-5/subscription/test-locally.mdx:168]:
```suggestion
This guide demonstrated how to test a VRF v2.5 consumer contract on your local blockchain. The guide uses the Remix IDE for learning purposes, but you can reuse the same [testing logic](#testing-logic) in another development environment, such as Hardhat. For example, see the Hardhat Starter Kit [RandomNumberConsumer unit tests](https://github.com/smartcontractkit/hardhat-starter-kit/blob/main/test/unit/RandomNumberConsumer.spec.js).
```

File: src/content/vrf/v2-5/subscription/test-locally.mdx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+   You can set:
+
+   - `_BASEFEE` to <CopyText text="100000000000000000" code/>
+   - `_GASPRICELINK` to <CopyText text="1000000000" code/>
+   - `_WEIPERUNITLINK` to the current LINK/ETH price. Click the "Latest Price" button to view it:
+     {" "}
+     <LatestPrice client:idle feedAddress={linkEth.link.eth.sepolia.address} supportedChain="ETHEREUM_SEPOLIA" />
+
+1. Click on _transact_ to deploy the _VRFCoordinatorV2_5Mock_ contract.
```
Comment (dwightjl) [inline src/content/vrf/v2-5/subscription/test-locally.mdx:80]:
```suggestion
1. Click _transact_ to deploy the _VRFCoordinatorV2_5Mock_ contract.
```

File: src/content/vrf/v2-5/subscription/test-locally.mdx [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+1. Once deployed, you should see the _VRFCoordinatorV2_5Mock_ contract under _Deployed Contracts_.
+
+   <ClickToZoom src="/images/vrf/v2-5/mock/deployed-mock.png" />
+
+1. Note the address of the deployed contract.
+
+### Create and fund a subscription
+
+1. Click on `createSubscription` to create a new subscription.
```
Comment (dwightjl) [inline src/content/vrf/v2-5/subscription/test-locally.mdx:90]:
```suggestion
1. Click `createSubscription` to create a new subscription.
```

File: src/content/vrf/v2-5/subscription/test-locally.mdx [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+   <ClickToZoom src="/images/vrf/v2-5/mock/deploy-consumer.png" />
+
+1. Under _DEPLOY_, fill in the following parameters:
+
+   - `SUBSCRIPTIONID` with your subscription ID
+   - `VRFCOORDINATOR` with the deployed _VRFCoordinatorV2_5Mock_ address
+   - `_KEYHASH_` with an arbitrary `bytes32` (In this example, you can set the _KEYHASH_ to <CopyText text="0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae" code/>).
+
+1. Click on _transact_ to deploy the _RandomNumberConsumerV2_5_ contract.
```
Comment (dwightjl) [inline src/content/vrf/v2-5/subscription/test-locally.mdx:112]:
```suggestion
1. Click _transact_ to deploy the _RandomNumberConsumerV2_5_ contract.
```

File: src/content/vrf/v2-5/subscription/test-locally.mdx [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+1. After the consumer contract is deployed, you should see the _RandomNumberConsumerV2_5_ contract under _Deployed Contracts_. Note the address of the deployed contract.
+
+### Add the consumer contract to your subscription
+
+1. Under _Deployed Contracts_, open the functions list of your deployed _VRFCoordinatorV2_5Mock_ contract.
+
+1. Click on _addConsumer_ and fill in the `_subid` with your subscription ID and `_consumer` with your deployed consumer contract address.
```
Comment (dwightjl) [inline src/content/vrf/v2-5/subscription/test-locally.mdx:120]:
```suggestion
1. Click _addConsumer_ and fill in the `_subid` with your subscription ID and `_consumer` with your deployed consumer contract address.
```

File: src/content/vrf/v2-5/subscription/test-locally.mdx [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+1. Click on `transact`.
+
+### Check the results
+
+1. Under _Deployed Contracts_, open the functions list of your deployed _RandomNumberConsumerV2_5_ contract.
+
+1. Each time you make a VRF request, your consumer contract requests two random words. After the request is fulfilled, the two random words are stored in the `s_randomWords` array. You can check the stored random words by reading the two first indexes of the `s_randomWords` array. To do so, click on the _s_randomWords_ function and:
+
+   1. Fill in the index with _0_ then click on _call_ to read the first random word.
```
Comment (dwightjl) [inline src/content/vrf/v2-5/subscription/test-locally.mdx:158]:
```suggestion
   1. Fill in the index with _0_ then click _call_ to read the first random word.
```

File: src/content/vrf/v2-5/subscription/test-locally.mdx [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+      {" "}
+
+      <ClickToZoom src="/images/vrf/v2-5/mock/show-random-word.png" />
+
+   1. You can read the second random word in a similar way: fill in the index with _1_ then click on _call_ to display the second random word.
```
Comment (dwightjl) [inline src/content/vrf/v2-5/subscription/test-locally.mdx:164]:
```suggestion
   1. You can read the second random word in a similar way: fill in the index with _1_ then click _call_ to display the second random word.
```

**Review by dwightjl — APPROVED** [review]
Minor edits

**Pull request merged**